### PR TITLE
[sumconnector] implement summing logic

### PR DESCRIPTION
**Pull request by greatestusername**
**Description:** 
- Adds connector and summing logic
- Adds testing of traces and spans summing

Note: testing and test data makes up the bulk of the lines in this PR

**Link to tracking Issue:** 32669

**Testing:** 
- condition, attribute, default attribute, multiple condition, multiple attributes, multiple metrics for traces / spans telemetry types

**Documentation:** No new docs

### CODE REVIEW

File: connector/sumconnector/sum.go
Comment (crobert-1) [inline connector/sumconnector/sum.go:56]: It may be worthwhile to include that the default is to convert to floats for strings in the README, but I'm not sure if there's a clear way to include it there.

**Review by crobert-1 — COMMENTED** [review]
If it's not too much trouble, could we split this PR into three: one for each of metrics, traces, and logs?

_This was a pretty minimal review, if we're able to break it up it would be really helpful in being able to give a more detailed review._

File: cmd/otelcontribcol/go.mod
Comment (atoulme) [inline cmd/otelcontribcol/go.mod:7]: just a good idea to not touch this, it will burn
```suggestion
toolchain go1.22.6
```

Comment (greatestusername) [reply cmd/otelcontribcol/go.mod:7]: Not even sure how this got changed. 🚀

Comment (greatestusername) [reply cmd/otelcontribcol/go.mod:7]: still getting failure even after running
`make genotelcontribcol` 😢 
https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/10722164772/job/29732593501?pr=34797

File: connector/sumconnector/sum.go
Comment (crobert-1) [inline connector/sumconnector/sum.go:136]: It'd be good to add a comment here about why this is necessary.

Comment (greatestusername) [reply connector/sumconnector/sum.go:137]: Added one for clarification this addresses float precision below 6 decimals.

Or would you like me to explicitly call out the testing and add a TODO: add float threshold comparison option to `CompareMetrics`? Just let me know and I'm happy to do that.

File: connector/sumconnector/connector.go
Comment (greatestusername) [reply connector/sumconnector/connector.go:104]: 👍 added

File: .chloggen/32669-sumconnector-sum-logic.yaml
Comment (greatestusername) [reply .chloggen/32669-sumconnector-sum-logic.yaml]: yeah good call. removed!

File: connector/sumconnector/sum.go
Comment (crobert-1) [reply connector/sumconnector/sum.go:137]: Is there a reason this is necessary in the functionality of the component, instead of just being used for tests? My original understanding was it would just be used for comparing test values, I didn't think we'd need this when setting metric values.

Comment (greatestusername) [reply connector/sumconnector/sum.go:137]: Oh agree!
This is primarily here so you can check out the logic and the tests can pass as is.

The issue as I see it is currently `CompareMetrics` func can't ignore float precision and to do that comparison with float precision threshold I'd need to basically duplicate the entire CompareMetrics functions (the nested parsing especially) and add the float precision thresholding function. That or put this all on hold to PR that into the main repo's testing options.

Unless there is some more creative way to do this comparing of two metric values that is eluding me (please say there is!)

Comment (greatestusername) [reply connector/sumconnector/sum.go:137]: Added an issue and will happily add an option to ignore tiny deviations of float values to `pmetrictest` once the issue has been triaged/assigned/etc. 👍

Comment (greatestusername) [reply connector/sumconnector/sum.go]: ok we're using the new option to ignore float precision after 3 decimals.

File: connector/sumconnector/README.md
Comment (crobert-1) [inline connector/sumconnector/README.md:61]:
```suggestion
- Telemetry type: Nested below the `sum:` connector declaration. Declared as `logs:` in the [Basic Example](#basic-configuration).
```

File: connector/sumconnector/README.md
Comment (crobert-1) [inline connector/sumconnector/README.md:69]:
```suggestion
- `attributes`: Declaration of attributes to include. Any of these attributes found will generate a separate sum for each set of unique combination of attribute values and output as its own datapoint in the metric time series.
```

File: connector/sumconnector/README.md
Comment (crobert-1) [inline connector/sumconnector/README.md:68]:
```suggestion
- `conditions`: [OTTL syntax](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/LANGUAGE.md) can be used to provide conditions for processing incoming telemetry. Conditions are ORed together, so if any condition is met the attribute's value will be included in the resulting sum.
```

**Review by crobert-1 — APPROVED** [review]
Just some whitespace nits, looks good though. Thanks for adding the pmetric ignore test condition too!

**Comment by greatestusername** [comment]
Thanks for the approval! README whitespace nits addressed.

**Comment by greatestusername** [comment]
@crobert-1 Is there anyone else I need approvals from or just waiting for someone to merge in the next round?
Thanks!

**Comment by crobert-1** [comment]
> Is there anyone else I need approvals from or just waiting for someone to merge in the next round?

We should be good, I've added `ready to merge`. 👍

**Pull request merged**
